### PR TITLE
Validate about when sampler is unnormalizedCoordinates.

### PR DESCRIPTION
**Pull request by locke-lunarg**
1) VUID-vkCmdDraw-None-02699
Descriptors in each bound descriptor set, specified via vkCmdBindDescriptorSets, must be valid if they are statically used by the VkPipeline bound to the pipeline bind point used by this command

Changed `kVUID_Core_DrawState_DescriptorSetNotUpdated `into `VUID-vkCmdDraw-None-02699.`

2) VUID-vkCmdDraw-None-02702
If the VkPipeline object bound to the pipeline bind point used by this command accesses a VkSampler object that uses unnormalized coordinates, that sampler must not be used to sample from any VkImage with a VkImageView of the type VK_IMAGE_VIEW_TYPE_3D, VK_IMAGE_VIEW_TYPE_CUBE, VK_IMAGE_VIEW_TYPE_1D_ARRAY, VK_IMAGE_VIEW_TYPE_2D_ARRAY or VK_IMAGE_VIEW_TYPE_CUBE_ARRAY, in any shader stage

Check if sampler is unnormalizedCoordinates, the imageview type could not be `VK_IMAGE_VIEW_TYPE_3D, VK_IMAGE_VIEW_TYPE_CUBE, VK_IMAGE_VIEW_TYPE_1D_ARRAY, VK_IMAGE_VIEW_TYPE_2D_ARRAY or VK_IMAGE_VIEW_TYPE_CUBE_ARRAY.`

If the descriptor type is `VK_DESCRIPTOR_TYPE_COMBINED_IMAGE_SAMPLER`, this is a easy case. But if the descriptor types are `VK_DESCRIPTOR_TYPE_SAMPLED_IMAGE and VK_DESCRIPTOR_TYPE_SAMPLER,` that is much complex. It needs to find out which sampler is used with image. In `interface_var`, it saves the sampler slots that are used with image. And then `DescriptorReqirement` have the pointer of the slots from `interface_var`. And then when validate ImageDescriptor, it will find out the sampler to verity it.

### CODE REVIEW

**Comment by locke-lunarg** [comment]
VUID-vkCmdDraw-None-02703
If the VkPipeline object bound to the pipeline bind point used by this command accesses a VkSampler object that uses unnormalized coordinates, that sampler must not be used with any of the SPIR-V OpImageSample* or OpImageSparseSample* instructions with ImplicitLod, Dref or Proj in their name, in any shader stage

Check if it has     
OpImageSampleImplicitLod, OpImageSampleDrefImplicitLod, OpImageSampleProjImplicitLod, OpImageSampleProjDrefImplicitLod, OpImageSparseSampleImplicitLod, OpImageSparseSampleDrefImplicitLod, OpImageSparseSampleProjImplicitLod, OpImageSparseSampleProjDrefImplicitLod in shader code.

File: layers/core_validation_types.h [hunk 1]
```diff
@@ -797,17 +821,24 @@ struct PIPELINE_LAYOUT_STATE : public BASE_NODE {
...
+
 // Shader typedefs needed to store StageStage below
 struct interface_var {
     uint32_t id;
     uint32_t type_id;
     uint32_t offset;
     int32_t input_index;  // index = -1 means that it's not input attachment.
+
+    std::vector<SamplerUsedByImage>
+        samplers_used_by_image;  // It's for VK_DESCRIPTOR_TYPE_SAMPLED_IMAGE to check it's used with VK_DESCRIPTOR_TYPE_SAMPLER.
```
Comment (Tony-LunarG) [inline layers/core_validation_types.h:833]: A better comment might be "List of samplers that sample a given image"

File: layers/core_validation_types.h [hunk 2]
```diff
@@ -797,17 +821,24 @@ struct PIPELINE_LAYOUT_STATE : public BASE_NODE {
...
+
     bool is_patch;
     bool is_block_member;
     bool is_relaxed_precision;
     bool is_writable;
     bool is_atomic_operation;
+    bool is_sampler_implicitLod;
```
Comment (Tony-LunarG) [inline layers/core_validation_types.h:840]: 02703 goes beyond just implicitLod.  For instance OpImageSampleDrefExplicitLod (has Dref), OpImageSampleProjExplicitLod (has Proj) and OpImageSampleProjDrefExplicitLod (has both)

File: layers/shader_validation.cpp [hunk 3]
```diff
@@ -858,106 +858,223 @@ static bool AtomicOperation(uint32_t opcode) {
...
+            }
+        }
+    }
+};
+
 // Check writable, image atomic operation
 static void IsSpecificDescriptorType(SHADER_MODULE_STATE const *module, const spirv_inst_iter &id_it, bool is_storage_buffer,
-                                     bool is_check_writable, interface_var &out_interface_var) {
+                                     bool is_check_writable, interface_var &out_interface_var,
+                                     shader_moudle_used_operators &used_operators) {
```
Comment (Tony-LunarG) [inline layers/shader_validation.cpp:1002]: moudle -> module

**Review by Tony-LunarG — COMMENTED** [review]
I think 02703 needs to check for more Op types

File: layers/core_validation_types.h (hunk 1 above)
Comment (locke-lunarg) [reply layers/core_validation_types.h]: fixed

File: layers/shader_validation.cpp (hunk 3 above)
Comment (locke-lunarg) [reply layers/shader_validation.cpp]: fixed

File: layers/core_validation_types.h (hunk 2 above)
Comment (locke-lunarg) [reply layers/core_validation_types.h]: Added OpImageSampleProjExplicitLod, OpImageSparseSampleProjExplicitLod, OpImageSampleDrefExplicitLod, OpImageSampleProjDrefExplicitLod,  OpImageSparseSampleDrefExplicitLod, OpImageSparseSampleProjDrefExplicitLod

**Review by Tony-LunarG — APPROVED** [review]
LGTM with the caveat that I don't review much shader validation code

File: layers/core_validation_error_enums.h [hunk 4]
```diff
@@ -57,7 +57,7 @@ static const char DECORATE_UNUSED *kVUID_Core_MemTrack_RebindObject = "UNASSIGNE
 static const char DECORATE_UNUSED *kVUID_Core_DrawState_ClearCmdBeforeDraw = "UNASSIGNED-CoreValidation-DrawState-ClearCmdBeforeDraw";
 static const char DECORATE_UNUSED *kVUID_Core_DrawState_CommandBufferSingleSubmitViolation = "UNASSIGNED-CoreValidation-DrawState-CommandBufferSingleSubmitViolation";
 static const char DECORATE_UNUSED *kVUID_Core_DrawState_DescriptorSetNotBound = "UNASSIGNED-CoreValidation-DrawState-DescriptorSetNotBound";
-static const char DECORATE_UNUSED *kVUID_Core_DrawState_DescriptorSetNotUpdated = "UNASSIGNED-CoreValidation-DrawState-DescriptorSetNotUpdated";
+// static const char DECORATE_UNUSED *kVUID_Core_DrawState_DescriptorSetNotUpdated = "UNASSIGNED-CoreValidation-DrawState-DescriptorSetNotUpdated";
```
Comment (ncesario-lunarg) [inline layers/core_validation_error_enums.h:60]: Should this be removed if no longer used?

Comment (locke-lunarg) [reply layers/core_validation_error_enums.h:60]: We prefer to leave it for tracing back through the git history.

Here is the information for Lobo.
https://github.com/KhronosGroup/Vulkan-ValidationLayers/pull/2085#discussion_r465989751

Comment (ncesario-lunarg) [reply layers/core_validation_error_enums.h:60]: 👍 thanks!

**Pull request merged**